### ISSUE TITLE
revise jxlsave to reduce memory use

### ISSUE THREAD

**Issue by tulzke** [libvips/libvips#3132]
## Bug report

When using libvips to convert to jxl, there is a huge memory consumption.
Yes, the image is large. It has a size of 23205x15802. But this is no more than 1GB in raw format.
It eats up all the memory, after which it is killed by the OS via SIGKILL. (My system uses 32Gb RAM)
At the same time, cjxl (the libjxl utility) does this without problems.

**To Reproduce**
Download that image: https://disk.yandex.ru/d/7YnSsGURImI7Vg
Use cli command: vips copy to_jpg_example.jpg converted.jxl

**Expected behavior**
Successful conversion to jxl

**Actual behavior**
Killing a process after uncontrolled memory consumption

**Screenshots**
![image](https://user-images.githubusercontent.com/62292054/199232636-46808e8b-09ec-4ae4-a721-135449c2cb69.png)

**Environment**
 - OS: Zorin OS 16.2 (based on Ubuntu 20.04)
 - Vips: vips-8.13.2
 - libjxl: v0.7.0

**Comment by jcupitt** [comment]
Hi @Tulzke,

libvips is still using the original one-shot encode mode for libjxl, so yes, memory consumption is high for large images. I tried with a largish image here:

```
$ vipsheader st-francis.jpg
st-francis.jpg: 30000x26319 uchar, 3 bands, srgb, jpegload
$ /usr/bin/time -f %M:%e vips copy st-francis.jpg st-francis.jxl
memory: high-water mark 2.46 GB
34650700:30.25
```

So 30gb and 30s for a 30k x 30k pixel image.

I think there's now a streaming encode mode, so yes, libvips should switch to that. As far as I know there's no streaming decode though :-(  There's talk of adding a tiled mode to JXL, which would be even better, of course.

Let's tag this as an enhancement and try to revise this for 8.14.

**Comment by jcupitt** [comment]
I tried your image and saw:

```
$ /usr/bin/time -f %M:%e vips copy ~/to_jxl_example.jpg x.jxl 
memory: high-water mark 1.22 GB
48146648:66.13
```

So 48gb of memory and 66s of real time. I wonder why it's more difficult to encode? Perhaps it's related to the colour profile.

**Comment by Tulzke** [comment]
Hello, @jcupitt.
Thanks for the reply. For my purposes (efficient storage of original images for the possibility of subsequent resizing) I went to the trick and reduced the too large images to 4k. In this case, I managed to achieve adequate resource consumption.

**Comment by atjn** [comment]
This is causing errors in production for me. I know the integration is still very early, and I wasn't expecting it to be flawless, but just wanted to let you know :)

I ran some tests with wasm-vips, and it seems like any image that is roughly larger than 3500x3500 will crash the process. This is a problem because my pipeline deals with images for 4K screens, which are just in that ballpark of resolution. Here is a test case you can run in the browser: [test case](https://wasm-vips.kleisauke.nl/playground/?modules=jxl&deflate=bZHLasMwEEV_ZfDGCgQ5bekmTbJIaSBQKPSxirPwY-zKlSUjy4-09N87so1LoRtJM3Pv0Wh08kIl0YIWJWyhFVXNj2WUI1fYHYwuD0Ii83UneVHl_uIuVKEKAngpIynRQEblGjptPqBoakuxwvWgcNCBSWRusBafyK5nv8PWEMW6RYiqyuiew83tatW7BTohpSsaS6z_QFerkRSqRCu6Vjd232QZNbQlLe-MsPiqxxTzedHLofVRHEsdk44eCHs6stPsPi_hC-ylwjX4wk0hcE74_ut9e34kO608MRhZfIoLTCzFzFV_taLMSZfqpClR2Un8INFFjPjjOGnntUlIObEpN3tytJNhfzmm9A-NrRrrLzjNDFV6_y5kyojgQN7S26SiBZFuQ28Uht5uE1BuN1S98w8)

Thank you for already having such good JPEG XL support!

**Comment by kleisauke** [comment]
It looks like libjxl is considering supporting a streaming/chunk-wise encode API, see:
https://github.com/libjxl/libjxl/pull/2632

**Comment by sandstrom** [comment]
There were some updates in 0.9, with better support for streaming in libjxl.

https://github.com/libjxl/libjxl/releases/tag/v0.9.0

Also related is this issue in libjxl's repo, where there seem to be some interest from the libvips team to contribute to libvips directly. I think that would be ideal!

https://github.com/libjxl/libjxl/issues/2640#issuecomment-1824831526

**Comment by JinEnMok** [comment]
Per Cloudinary's [recent test](https://cloudinary.com/blog/jpeg-xl-and-the-pareto-front#lossless_huge_improvements_in_memory_usage_and_speed), an 80 (?) MP image now requires only ~1GB of memory as opposed to 8GB.

Compressing OP's file also took only 1GB (per VIPS's 'highwater mark') with `libvips` 8.15.1 and `libjxl` 0.10.1, and although Gwenview refused to display the file, `djxl` was able to decode it to ppm and it displayed correctly.

**Comment by jcupitt** [comment]
Yes, it's a nice improvement! We've started building libvips binaries with 0.10.1, eg.:

https://github.com/jcupitt/vipsdisp/releases/tag/v3.0.4

**Comment by jcupitt** [comment]
> per VIPS's 'highwater mark'

The libvips highwater mark just counts libvips pixel buffers, it won't include memory allocated by external libraries. I usually test memuse with eg.:

```
$ vipsheader nina.jpg
nina.jpg: 6048x4032 uchar, 3 bands, srgb, jpegload
$ /usr/bin/time -f %M:%e vips copy nina.jpg x.jxl
memory: high-water mark 74.22 MB
2303124:4.46
$ /usr/bin/time -f %M:%e vips copy nina.jpg x.jxl
memory: high-water mark 74.20 MB
463988:1.00
```

So that's about a 5x improvement in encoding speed and memory use with 0.10.1.
